src/stan/services/optimize/bfgs.hpp

```diff
@@ -0,0 +1,200 @@
+       *
+       * @param model Input model to test (with data already instantiated)
+       * @param init var context for initialization
+       * @param random_seed random seed for the pseudo random number generator
+       * @param chain chain id to advance the pseudo random number generator
+       * @param init_radius radius to initialize
+       * @param[in] init_alpha
+       * @param[in] tol_obj
+       * @param[in] tol_rel_obj
+       * @param[in] tol_grad
+       * @param[in] tol_rel_grad
+       * @param[in] tol_param
+       * @param[in] num_iterations maximum number of iterations
+       * @param[in] save_iterations indicates whether all the interations should
+       *   be saved to the parameter_writer
+       * @param[in] refresh how often to write output to message_writer
+       * @param[out] interrupt interrupt callback to be called every iteration
+       * @param[out] message_writer output for messages
+       * @param[out] init_writer Writer callback for unconstrained inits
+       * @param[out] parameter_writer output for parameter values
+       *
+       * @return stan::services::error_codes::OK (0) if successful
+       */
+      template <class Model>
+      int bfgs(Model& model,
+               stan::io::var_context& init,
+               unsigned int random_seed,
+               unsigned int chain,
+               double init_radius,
+               double init_alpha,
+               double tol_obj,
+               double tol_rel_obj,
+               double tol_grad,
+               double tol_rel_grad,
+               double tol_param,
+               int num_iterations,
+               bool save_iterations,
+               int refresh,
+               callbacks::interrupt& interrupt,
+               callbacks::writer& message_writer,
+               callbacks::writer& init_writer,
+               callbacks::writer& parameter_writer) {
+        boost::ecuyer1988 rng = stan::services::util::rng(random_seed, chain);
+
+        std::vector<int> disc_vector;
+        std::vector<double> cont_vector
+          = stan::services::util::initialize(model, init, rng, init_radius,
+                                             false,
+                                             message_writer, init_writer);
+
+        std::stringstream bfgs_ss;
+        typedef stan::optimization::BFGSLineSearch
+          <Model, stan::optimization::BFGSUpdate_HInv<> > Optimizer;
+        Optimizer bfgs(model, cont_vector, disc_vector, &bfgs_ss);
+        bfgs._ls_opts.alpha0 = init_alpha;
+        bfgs._conv_opts.tolAbsF = tol_obj;
+        bfgs._conv_opts.tolRelF = tol_rel_obj;
+        bfgs._conv_opts.tolAbsGrad = tol_grad;
+        bfgs._conv_opts.tolRelGrad = tol_rel_grad;
+        bfgs._conv_opts.tolAbsX = tol_param;
+        bfgs._conv_opts.maxIts = num_iterations;
+
+        double lp = bfgs.logp();
+
+        std::stringstream msg;
+        msg << "Initial log joint probability = " << lp;
+        message_writer(msg.str());
```

Comment:
Can't change it to that: there is no method for accepting double values and it'll change the spacing to something that's different. I left the output identical to the old output to be safe.

The repeated code isn't new code. This is code that's moved that GitHub isn't tracking so well as moved code.